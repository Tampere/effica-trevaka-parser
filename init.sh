#!/bin/bash

# SPDX-FileCopyrightText: 2023-2024 Tampere region
#
# SPDX-License-Identifier: LGPL-2.1-or-later

set -e
set -u

function create_database() {
  psql -v ON_ERROR_STOP=1 --username "$POSTGRES_USER" <<-EOSQL
    CREATE DATABASE $1;
    \connect $1
    CREATE SCHEMA migration;
EOSQL
}

create_database vesilahti
