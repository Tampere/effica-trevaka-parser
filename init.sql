-- SPDX-FileCopyrightText: 2021 City of Tampere
--
-- SPDX-License-Identifier: LGPL-2.1-or-later

-- Add extensions here as needed (for UUIDs at least)
CREATE SCHEMA IF NOT EXISTS ext;
CREATE SCHEMA IF NOT EXISTS migration;

CREATE EXTENSION IF NOT EXISTS "uuid-ossp" WITH SCHEMA ext;
CREATE EXTENSION IF NOT EXISTS btree_gist WITH SCHEMA ext;