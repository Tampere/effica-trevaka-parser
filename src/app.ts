import express from "express"
import importApi from "./api/import"
import checkApi from "./api/check"

const app: express.Application = express()

app.use(express.json());

app.use("/import", importApi)
app.use("/check", checkApi)

export default app