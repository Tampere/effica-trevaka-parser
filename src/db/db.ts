import pgPromise from "pg-promise"
import { config } from "../config"

export const pgp = pgPromise({})

// date parsing seems to be scuffed, does local tz transformation but declares it UTC -> corrupts dates
//pgp.pg.types.setTypeParser(1082, dateParser)

const migrationDb = pgp({ ...config.migrationDb })

export default migrationDb