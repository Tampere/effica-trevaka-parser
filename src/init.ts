// SPDX-FileCopyrightText: 2021 City of Tampere
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import migrationDb from "./db/db";
import { baseQueryParameters, runQueryFile } from "./util/queryTools";

export const initDb = async () => {
    return await migrationDb.tx(async (t) => {
        await runQueryFile("init.sql", t, baseQueryParameters);
        await runQueryFile("functions.sql", t, baseQueryParameters);
    });
};
