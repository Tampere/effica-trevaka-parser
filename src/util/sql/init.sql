-- SPDX-FileCopyrightText: 2021 City of Tampere
--
-- SPDX-License-Identifier: LGPL-2.1-or-later

CREATE TABLE IF NOT EXISTS ${migrationSchema:name}.idmap (
    type TEXT NOT NULL,
    effica_guid TEXT NOT NULL,
    evaka_id UUID NOT NULL,
    created TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
    updated TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
    PRIMARY KEY (type, effica_guid)
);
