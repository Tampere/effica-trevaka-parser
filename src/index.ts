// SPDX-FileCopyrightText: 2021 City of Tampere
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import app from "./app"
import { config } from "./config"

app.listen(config.port, () => {
    console.log(`Parser app listening at http://localhost:${config.port}`)
})

