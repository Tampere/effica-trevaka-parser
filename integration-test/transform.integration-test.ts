// SPDX-FileCopyrightText: 2021 City of Tampere
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import request from "supertest"
import app from "../src/app"
import db from "../src/db/db"
import { initDb } from "../src/init"
import { dropTable, truncateEvakaTables } from "../src/util/queryTools"
import { setupTable, setupTransfers, setupTransformations } from "../src/util/testTools"

const baseUrl = "/transform"
let cleanUps: string[] = []

let evakaDataCleanups: string[] = [
    "person",
]

//order based on dependency
const baseDataTables =
    [
        "persons",
        "codes",
        "families_exclusion",
        "families",
        "specialneeds",
        "specialmeans",
        "units",
        "departments",
        "extentmap",
        "unitmap",
        "childmindermap",
        "placements_exclusion",
        "placements",
        "placementextents",
        "feedeviations",
        "income",
        "incomerows",
        "decisions",
        "paydecisions",
        "paydecisionrows",
        "applications",
        "applicationrows",
        "dailyjournals",
        "dailyjournalrows",
        "timestampheaders",
        "timestampdetails",
        "archiveddocument",
        "evaka_areas",
        "evaka_daycare",
        "daycare_oid_map"

    ]


const personExpectation = {
    id: expect.any(String),
    date_of_birth: expect.any(String)
}

beforeAll(async () => {
    await initDb()

    for (const table of baseDataTables) {
        const importTarget = table === "archiveddocument" ? "archiveddocument" : undefined
        await setupTable(table, importTarget)
    }
})

beforeEach(() => {
})
afterEach(async () => {
    for (const table of cleanUps) {
        await dropTable(table)
    }
    cleanUps = []

    await truncateEvakaTables(evakaDataCleanups)

})

afterAll(async () => {
    try {
        for (const table of baseDataTables) {
            await dropTable(table)
        }
    } finally {
        return await db.$pool.end()
    }
})

describe("GET /transform positive", () => {
    it("should return transformed persons", async () => {
        cleanUps = ["evaka_person"]
        await positiveTransformSnapshotTest(
            "persons",
            Array(10).fill(personExpectation)
        )
    })
    it("should return transformed families", async () => {
        cleanUps = ["evaka_person", "evaka_fridge_child", "evaka_fridge_partner"]
        await setupTransformations(["persons"])

        const fridgeChildExpectation =
            [
                {
                    id: expect.any(String),
                    head_of_family: expect.any(String),
                    child_id: expect.any(String),
                    start_date: expect.any(String)
                },
                {
                    id: expect.any(String),
                    head_of_family: expect.any(String),
                    child_id: expect.any(String),
                    start_date: expect.any(String)
                },
                {
                    id: expect.any(String),
                    head_of_family: expect.any(String),
                    child_id: expect.any(String),
                    start_date: expect.any(String)
                }
            ]

        const fridgePartnerExpectation =
            [
                {
                    partnership_id: expect.any(String),
                    person_id: expect.any(String),
                    start_date: expect.any(String)
                },
                {
                    partnership_id: expect.any(String),
                    person_id: expect.any(String),
                    start_date: expect.any(String),
                },
                {
                    partnership_id: expect.any(String),
                    person_id: expect.any(String),
                    start_date: expect.any(String),
                },
                {
                    partnership_id: expect.any(String),
                    person_id: expect.any(String),
                    start_date: expect.any(String),
                }
            ]

        await positiveTransformSnapshotTest(
            "families",
            { child: fridgeChildExpectation, partner: fridgePartnerExpectation }
        )

    })
    it("should return transformed special needs", async () => {
        await setupTransformations(["persons"])

        const assistanceNeedExpectation = {
            id: expect.any(String),
            child_id: expect.any(String),
        }

        await positiveTransformSnapshotTest(
            "special_needs",
            {
                "assistanceNeeds": Array(1).fill(assistanceNeedExpectation),
                "assistanceNeedsTodo": Array(0).fill(assistanceNeedExpectation),
            }
        )
    })
    it("should return transformed special means", async () => {
        await setupTransformations(["persons"])

        const assistanceActionExpectation = {
            id: expect.any(String),
            child_id: expect.any(String),
        }

        await positiveTransformSnapshotTest(
            "special_means",
            {
                assistanceActions: Array(1).fill(assistanceActionExpectation),
                assistanceActionsTodo: Array(0).fill(assistanceActionExpectation),
            }
        )
    })
    it("should return transformed departments", async () => {
        cleanUps = ["evaka_daycare_group"]

        const daycareGroupExpectation =
            [
                {
                    id: expect.any(String),
                    start_date: expect.any(String),
                    end_date: null
                },
                {
                    id: expect.any(String),
                    start_date: expect.any(String),
                    end_date: null
                }
            ]

        await positiveTransformSnapshotTest(
            "departments",
            daycareGroupExpectation
        )
    })

    it("should return transformed placements", async () => {
        await setupTransformations(["persons", "families", "departments", "application"])

        const placementExpectation = {
            id: expect.any(String),
            child_id: expect.any(String),
            daycare_group_id: expect.any(String)
        }
        const applicationExpectation = {
            id: expect.any(String),
            child_id: expect.any(String),
            guardian_id: expect.any(String),
        }

        await positiveTransformSnapshotTest(
            "placements",
            {
                placements: [
                    placementExpectation, // p=11, e=111
                    placementExpectation, // p=12, e=12
                    { ...placementExpectation, daycare_group_id: null }, // p=13, e=131
                    placementExpectation, // p=18, e=18
                    placementExpectation, // p=20, e=200
                    placementExpectation, // p=21, e=210
                    placementExpectation, // p=22, e=220
                    placementExpectation, // p=23, e=230
                ],
                placementsTodo: [
                    placementExpectation, // p=11, p=112
                    { ...placementExpectation, daycare_group_id: null }, // p=14, e=14
                    { ...placementExpectation, daycare_group_id: null }, // p=15, e=15
                    { ...placementExpectation, daycare_group_id: null }, // p=17, e=17
                    { ...placementExpectation, daycare_group_id: null }, // p=17, e=17
                    placementExpectation, // p=19, e=191
                    placementExpectation, // p=19, e=192
                ],
                applications: [
                    applicationExpectation,
                    applicationExpectation,
                ]
            }
        )
    })

    it("should return transformed feedeviations", async () => {
        await setupTransformations(["persons", "departments", "placements"])

        const feeAlterationExpectation = {
            id: expect.any(String),
            person_id: expect.any(String)
        }

        await positiveTransformSnapshotTest(
            "feedeviations",
            Array(1).fill(feeAlterationExpectation)
        )
    })

    it("should return transformed income", async () => {
        cleanUps = ["evaka_income", "evaka_person"]
        await setupTransformations(["persons"])

        const incomeExpectation = {
            id: expect.any(String),
            person_id: expect.any(String),
        }

        await positiveTransformSnapshotTest(
            "income",
            Array(2).fill(incomeExpectation)
        )
    })

    it("should return transformed voucher value decisions", async () => {
        await setupTransformations(["persons", "families", "departments", "placements", "income"])

        const voucherValueDecisionExpectation = {
            id: expect.any(String),
            head_of_family_id: expect.any(String),
            partner_id: expect.any(String),
            child_id: expect.any(String),
        }

        await positiveTransformSnapshotTest(
            "voucher_value_decisions",
            {
                decisions: Array(1).fill(voucherValueDecisionExpectation),
                decisionsTodo: Array(0).fill(voucherValueDecisionExpectation)
            }
        )
    })

    it("should return transformed pay decisions", async () => {
        await setupTransformations(["persons", "departments", "placements"])

        const feeDecisionExpectation = {
            id: expect.any(String),
            head_of_family_id: expect.any(String),
            partner_id: expect.any(String) || null
        }
        const feeDecisionChildExpectation = {
            id: expect.any(String),
            fee_decision_id: expect.any(String),
            child_id: expect.any(String)
        }

        await positiveTransformSnapshotTest(
            "pay_decisions",
            {
                feeDecisions: [
                    { ...feeDecisionExpectation, partner_id: null },
                    feeDecisionExpectation,
                    { ...feeDecisionExpectation, partner_id: null }
                ],
                feeDecisionsTodo: [
                    { ...feeDecisionExpectation, partner_id: null }
                ],
                feeDecisionChildren: Array(3).fill(feeDecisionChildExpectation),
                feeDecisionChildrenTodo: Array(0).fill(feeDecisionChildExpectation),
            }
        )
    })

    it("should return transformed daily journals", async () => {
        await setupTransformations(["persons", "departments", "placements"])

        const absenceExpectation = {
            id: expect.any(String),
            child_id: expect.any(String),
        }

        const backupCareExpectation = {
            id: expect.any(String),
            child_id: expect.any(String),
            unit_id: expect.any(String),
            group_id: expect.any(String),
        }

        await positiveTransformSnapshotTest(
            "daily_journals",
            {
                absences: Array(1).fill(absenceExpectation),
                absencesTodo: Array(1).fill(absenceExpectation),
                backupCares: Array(2).fill(backupCareExpectation),
                backupCaresTodo: Array(0).fill(backupCareExpectation),
            }
        )
    })

    it("should return transformed timestamps", async () => {
        await setupTransformations(["persons"])

        const childAttendanceExpectation = {
            id: expect.any(String),
            child_id: expect.any(String),
        }

        await positiveTransformSnapshotTest(
            "timestamps",
            {
                childAttendances: Array(1).fill(childAttendanceExpectation),
                childAttendancesTodo: Array(0).fill(childAttendanceExpectation),
            }
        )
    })

    it("should return transformed pedagogical documents", async () => {
        await setupTransformations(["persons"])

        const pedagogicalDocumentExpectation = {
            id: expect.any(String),
            child_id: expect.any(String),
            attachment_id: expect.any(String),
        }

        await positiveTransformSnapshotTest(
            "pedagogical_documents",
            {
                data: Array(2).fill(pedagogicalDocumentExpectation),
                todo: Array(0).fill(pedagogicalDocumentExpectation),
            }
        )
    })

    it("should return transformed application", async () => {
        cleanUps = ["evaka_person", "evaka_fridge_child", "evaka_fridge_partner"]

        await setupTransformations(["persons", "families"])
        await setupTransfers(["persons", "families"])

        const applicationExpectation = {
            id: expect.any(String),
            child_id: expect.any(String),
            guardian_id: expect.any(String),
        }
        const applicationFormExpectation = {
            application_id: expect.any(String),
        }

        await positiveTransformSnapshotTest("application", {
            applications: Array(2).fill(applicationExpectation),
            applicationsTodo: Array(0).fill(applicationExpectation),
            applicationForms: Array(2).fill(applicationFormExpectation),
            applicationFormsTodo: Array(0).fill(applicationFormExpectation),
        })
    })

    it("should return cleanups", async () => {
        await setupTransformations(["persons", "departments", "placements"])

        const daycareGroupExpectation = {
            id: expect.any(String)
        }

        await positiveTransformSnapshotTest(
            "cleanup",
            {
                cleanedDaycareGroups: Array(1).fill(daycareGroupExpectation),
            }
        )
    })
})

const positiveTransformSnapshotTest = async (tableName: string, resultPattern?: any) => {
    const response = await transform(tableName)
    if (resultPattern) {
        expect(response.body).toMatchSnapshot(resultPattern)
    }
    else {
        expect(response.body).toMatchSnapshot()
    }
    return response
}

const transform = async (tableName: string) => {
    const queryObject = {
        returnAll: "true"
    }

    const url = `${baseUrl}/${tableName}`
    const response = await request(app).get(url).query(queryObject)
    expect(response.status).toBe(200)
    return response
}
