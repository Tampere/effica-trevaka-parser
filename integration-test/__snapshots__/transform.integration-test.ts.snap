// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /transform positive should return cleanups 1`] = `
Object {
  "cleanedDaycareGroups": Array [
    Object {
      "daycare_id": "19fec1fa-e2f1-11eb-8473-eb1f7ce94b07",
      "effica_id": 1251,
      "end_date": null,
      "id": Any<String>,
      "name": "Tujun tupajumit",
      "start_date": "2020-03-01T00:00:00.000Z",
    },
  ],
}
`;

exports[`GET /transform positive should return transformed application 1`] = `
Object {
  "applicationForms": Array [
    Object {
      "application_id": Any<String>,
      "effica_application_id": 12345,
      "effica_priority": 1,
      "preferred_start_date": "2021-09-01T00:00:00.000Z",
      "service_need_option_id": "86ef70a0-bf85-11eb-91e6-1fb57a101161",
      "unit_id": "19fec146-e2f1-11eb-8473-db55258254c5",
    },
  ],
  "applicationFormsTodo": Array [],
  "applications": Array [
    Object {
      "child_id": Any<String>,
      "duedate": "2021-11-29T00:00:00.000Z",
      "effica_guid": "12F28AA0-E059-4203-98F9-EE3D37738B99",
      "effica_id": 12345,
      "guardian_id": Any<String>,
      "id": Any<String>,
      "sentdate": "2021-08-01T00:00:00.000Z",
      "status": "SENT",
      "transferapplication": false,
      "type": "DAYCARE",
    },
  ],
  "applicationsTodo": Array [],
}
`;

exports[`GET /transform positive should return transformed daily journals 1`] = `
Object {
  "absences": Array [
    Object {
      "absence_type": "SICKLEAVE",
      "care_type": "DAYCARE",
      "child_id": Any<String>,
      "date": "2020-03-09T00:00:00.000Z",
      "effica_dailyjournalids": Array [
        123456,
      ],
      "id": Any<String>,
    },
  ],
  "absencesTodo": Array [
    Object {
      "absence_type": null,
      "care_type": "DAYCARE",
      "child_id": Any<String>,
      "date": "2020-03-24T00:00:00.000Z",
      "effica_dailyjournalids": Array [
        123456,
      ],
      "id": Any<String>,
      "reason": "TYPE MISSING",
    },
  ],
  "backupCares": Array [
    Object {
      "child_id": Any<String>,
      "end_date": "2020-03-17T00:00:00.000Z",
      "group_id": Any<String>,
      "id": Any<String>,
      "start_date": "2020-03-15T00:00:00.000Z",
      "unit_id": Any<String>,
    },
    Object {
      "child_id": Any<String>,
      "end_date": "2020-03-19T00:00:00.000Z",
      "group_id": Any<String>,
      "id": Any<String>,
      "start_date": "2020-03-19T00:00:00.000Z",
      "unit_id": Any<String>,
    },
  ],
  "backupCaresTodo": Array [],
}
`;

exports[`GET /transform positive should return transformed departments 1`] = `
Array [
  Object {
    "daycare_id": "19fec146-e2f1-11eb-8473-db55258254c5",
    "effica_id": 1231,
    "end_date": null,
    "id": Any<String>,
    "name": "Tenhon terhot",
    "start_date": Any<String>,
  },
  Object {
    "daycare_id": "19fec1fa-e2f1-11eb-8473-eb1f7ce94b07",
    "effica_id": 1251,
    "end_date": null,
    "id": Any<String>,
    "name": "Tujun tupajumit",
    "start_date": Any<String>,
  },
]
`;

exports[`GET /transform positive should return transformed families 1`] = `
Object {
  "child": Array [
    Object {
      "child_id": Any<String>,
      "child_ssn": "251118A9509",
      "end_date": "2020-11-30T00:00:00.000Z",
      "family_number": 1,
      "head_of_family": Any<String>,
      "hof_ssn": "020967-9992",
      "id": Any<String>,
      "start_date": Any<String>,
    },
    Object {
      "child_id": Any<String>,
      "child_ssn": "010619A9501",
      "end_date": "2037-05-31T00:00:00.000Z",
      "family_number": 2,
      "head_of_family": Any<String>,
      "hof_ssn": "020967-9992",
      "id": Any<String>,
      "start_date": Any<String>,
    },
    Object {
      "child_id": Any<String>,
      "child_ssn": "251118A9509",
      "end_date": "2036-11-24T00:00:00.000Z",
      "family_number": 2,
      "head_of_family": Any<String>,
      "hof_ssn": "020967-9992",
      "id": Any<String>,
      "start_date": Any<String>,
    },
  ],
  "partner": Array [
    Object {
      "effica_ssn": "020967-9992",
      "end_date": "2020-11-30T00:00:00.000Z",
      "family_number": 1,
      "indx": 1,
      "partnership_id": Any<String>,
      "person_id": Any<String>,
      "start_date": Any<String>,
    },
    Object {
      "effica_ssn": "180476-998M",
      "end_date": "2020-11-30T00:00:00.000Z",
      "family_number": 1,
      "indx": 2,
      "partnership_id": Any<String>,
      "person_id": Any<String>,
      "start_date": Any<String>,
    },
    Object {
      "effica_ssn": "020967-9992",
      "end_date": null,
      "family_number": 2,
      "indx": 1,
      "partnership_id": Any<String>,
      "person_id": Any<String>,
      "start_date": Any<String>,
    },
    Object {
      "effica_ssn": "180476-998M",
      "end_date": null,
      "family_number": 2,
      "indx": 2,
      "partnership_id": Any<String>,
      "person_id": Any<String>,
      "start_date": Any<String>,
    },
  ],
}
`;

exports[`GET /transform positive should return transformed feedeviations 1`] = `
Array [
  Object {
    "amount": 10,
    "id": Any<String>,
    "is_absolute": true,
    "notes": "ALENNUS AP/IP",
    "person_id": Any<String>,
    "type": "DISCOUNT",
    "valid_from": "2020-03-01T00:00:00.000Z",
    "valid_to": null,
  },
]
`;

exports[`GET /transform positive should return transformed income 1`] = `
Array [
  Object {
    "application_id": null,
    "data": Object {
      "DAILY_ALLOWANCE": Object {
        "amount": 1000,
        "coefficient": "DAILY_ALLOWANCE_21_5",
      },
      "MAIN_INCOME": Object {
        "amount": 450060,
        "coefficient": "MONTHLY_NO_HOLIDAY_BONUS",
      },
      "OTHER_INCOME": Object {
        "amount": 56000,
        "coefficient": "MONTHLY_NO_HOLIDAY_BONUS",
      },
      "PAID_ALIMONY": Object {
        "amount": 21500,
        "coefficient": "MONTHLY_NO_HOLIDAY_BONUS",
      },
    },
    "effect": "INCOME",
    "id": Any<String>,
    "income_total": 506060,
    "is_entrepreneur": false,
    "notes": "",
    "person_id": Any<String>,
    "person_ssn": "020967-9992",
    "valid_from": Any<String>,
    "valid_to": null,
  },
  Object {
    "application_id": null,
    "data": Object {
      "MAIN_INCOME": Object {
        "amount": 5000000,
        "coefficient": "MONTHLY_NO_HOLIDAY_BONUS",
      },
    },
    "effect": "INCOME",
    "id": Any<String>,
    "income_total": 5000000,
    "is_entrepreneur": false,
    "notes": "",
    "person_id": Any<String>,
    "person_ssn": "180476-998M",
    "valid_from": Any<String>,
    "valid_to": Any<String>,
  },
]
`;

exports[`GET /transform positive should return transformed pay decisions 1`] = `
Object {
  "feeDecisionChildren": Array [
    Object {
      "base_fee": 49200,
      "child_date_of_birth": "2019-06-01T00:00:00.000Z",
      "child_id": Any<String>,
      "effica_guid": "paydecisionrow38437813",
      "fee": 49200,
      "fee_decision_id": Any<String>,
      "final_fee": 49200,
      "id": Any<String>,
      "placement_unit_id": "19fec146-e2f1-11eb-8473-db55258254c5",
      "service_need_option_id": "86ef70a0-bf85-11eb-91e6-1fb57a101161",
      "sibling_discount": 0,
    },
  ],
  "feeDecisionChildrenTodo": Array [],
  "feeDecisions": Array [
    Object {
      "decision_number": "123456790",
      "decision_type": "NORMAL",
      "effica_guid": "paydecision3843781",
      "effica_ssn": "020967-9992",
      "end_date": null,
      "family_size": 2,
      "head_of_family_id": Any<String>,
      "head_of_family_income": Object {
        "data": Object {
          "MAIN_INCOME": 10020,
        },
        "effect": "INCOME",
        "total": 10020,
        "totalExpenses": 0,
        "totalIncome": 10020,
        "validFrom": null,
        "validTo": null,
        "worksAtECHA": false,
      },
      "id": Any<String>,
      "partner_id": Any<String>,
      "partner_income": Object {
        "data": Object {},
        "effect": "MAX_FEE_ACCEPTED",
        "total": 0,
        "totalExpenses": 0,
        "totalIncome": 0,
        "validFrom": null,
        "validTo": null,
        "worksAtECHA": false,
      },
      "start_date": "2020-03-01T00:00:00.000Z",
      "status": "SENT",
      "status_group": "SENT",
    },
  ],
  "feeDecisionsTodo": Array [],
}
`;

exports[`GET /transform positive should return transformed persons 1`] = `
Array [
  Object {
    "backup_phone": "",
    "date_of_birth": Any<String>,
    "effica_guid": "{AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAaa1}",
    "effica_ssn": "020967-9992",
    "email": "fake@notmaildomain.notext",
    "first_name": "Aimo",
    "id": Any<String>,
    "language": "fi",
    "last_name": "Authentic",
    "nationalities": Array [],
    "phone": "000-0000000",
    "post_office": "Yp채j채",
    "postal_code": "00001",
    "restricted_details_enabled": false,
    "social_security_number": "020967-9992",
    "source_system": "effica",
    "street_address": "Fakestreet 123",
  },
  Object {
    "backup_phone": "",
    "date_of_birth": Any<String>,
    "effica_guid": "{AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAaa2}",
    "effica_ssn": "020288-TP01",
    "email": "fake2@notmaildomain.notext",
    "first_name": "Teppo",
    "id": Any<String>,
    "language": "fi",
    "last_name": "Temporary",
    "nationalities": Array [],
    "phone": "111-1111111",
    "post_office": "Yp채j채",
    "postal_code": "00002",
    "restricted_details_enabled": false,
    "social_security_number": null,
    "source_system": "effica",
    "street_address": "StreetStreet 321",
  },
  Object {
    "backup_phone": "",
    "date_of_birth": Any<String>,
    "effica_guid": "{AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAaa3}",
    "effica_ssn": "180476-998M",
    "email": "fake3@notmaildomain.notext",
    "first_name": "Hilda",
    "id": Any<String>,
    "language": "fi",
    "last_name": "Hidden",
    "nationalities": Array [],
    "phone": "0222222222",
    "post_office": "",
    "postal_code": "",
    "restricted_details_enabled": true,
    "social_security_number": "180476-998M",
    "source_system": "effica",
    "street_address": "",
  },
  Object {
    "backup_phone": "222-2222223",
    "date_of_birth": Any<String>,
    "effica_guid": "{AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAaa4}",
    "effica_ssn": "010619A9501",
    "email": "fake3@notmaildomain.notext",
    "first_name": "Christian",
    "id": Any<String>,
    "language": "fi",
    "last_name": "Child",
    "nationalities": Array [],
    "phone": "222-2222222",
    "post_office": "Yp채j채",
    "postal_code": "00001",
    "restricted_details_enabled": false,
    "social_security_number": "010619A9501",
    "source_system": "effica",
    "street_address": "HideStreet 2",
  },
  Object {
    "backup_phone": "222-2222223",
    "date_of_birth": Any<String>,
    "effica_guid": "{AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAaa5}",
    "effica_ssn": "251118A9509",
    "email": "fake3@notmaildomain.notext",
    "first_name": "Camilla",
    "id": Any<String>,
    "language": "fi",
    "last_name": "Child",
    "nationalities": Array [],
    "phone": "222-2222222",
    "post_office": "Tenhola",
    "postal_code": "12000",
    "restricted_details_enabled": false,
    "social_security_number": "251118A9509",
    "source_system": "effica",
    "street_address": "HideStreet 2",
  },
  Object {
    "backup_phone": "",
    "date_of_birth": Any<String>,
    "effica_guid": "{AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAaa6}",
    "effica_ssn": "181101A9092",
    "email": null,
    "first_name": "",
    "id": Any<String>,
    "language": "fi",
    "last_name": "",
    "nationalities": Array [],
    "phone": "",
    "post_office": "",
    "postal_code": "",
    "restricted_details_enabled": false,
    "social_security_number": "181101A9092",
    "source_system": "effica",
    "street_address": "",
  },
]
`;

exports[`GET /transform positive should return transformed placements 1`] = `
Object {
  "placements": Array [
    Object {
      "child_id": Any<String>,
      "daycare_group_id": Any<String>,
      "effica_placement_nbr": 11,
      "effica_ssn": "010619A9501",
      "end_date": "2021-02-28T00:00:00.000Z",
      "id": Any<String>,
      "start_date": "2020-03-01T00:00:00.000Z",
      "type": "DAYCARE",
      "unit_id": "19fec146-e2f1-11eb-8473-db55258254c5",
    },
    Object {
      "child_id": Any<String>,
      "daycare_group_id": Any<String>,
      "effica_placement_nbr": 12,
      "effica_ssn": "010619A9501",
      "end_date": "2025-07-31T00:00:00.000Z",
      "id": Any<String>,
      "start_date": "2021-03-01T00:00:00.000Z",
      "type": "DAYCARE",
      "unit_id": "19fec146-e2f1-11eb-8473-db55258254c5",
    },
    Object {
      "child_id": Any<String>,
      "daycare_group_id": null,
      "effica_placement_nbr": 13,
      "effica_ssn": "251118A9509",
      "end_date": "2020-12-31T00:00:00.000Z",
      "id": Any<String>,
      "start_date": "2020-03-01T00:00:00.000Z",
      "type": "DAYCARE_PART_TIME",
      "unit_id": "19fec1fa-e2f1-11eb-8473-eb1f7ce94b07",
    },
  ],
  "placementsTodo": Array [
    Object {
      "child_id": Any<String>,
      "daycare_group_id": null,
      "effica_placement_nbr": 14,
      "effica_ssn": "251118A9509",
      "end_date": "2021-12-31T00:00:00.000Z",
      "id": Any<String>,
      "reason": "OVERLAPPING PLACEMENT",
      "start_date": "2021-01-01T00:00:00.000Z",
      "type": "DAYCARE",
      "unit_id": "19fec1fa-e2f1-11eb-8473-eb1f7ce94b07",
    },
    Object {
      "child_id": Any<String>,
      "daycare_group_id": null,
      "effica_placement_nbr": 15,
      "effica_ssn": "251118A9509",
      "end_date": "2024-07-31T00:00:00.000Z",
      "id": Any<String>,
      "reason": "OVERLAPPING PLACEMENT",
      "start_date": "2021-08-01T00:00:00.000Z",
      "type": "DAYCARE",
      "unit_id": "19fec1fa-e2f1-11eb-8473-eb1f7ce94b07",
    },
  ],
  "serviceNeeds": Array [
    Object {
      "effica_extent_code": "100",
      "effica_extent_nbr": 1,
      "effica_placement_nbr": 11,
      "end_date": "2021-02-28T00:00:00.000Z",
      "id": Any<String>,
      "option_id": "86ef70a0-bf85-11eb-91e6-1fb57a101161",
      "placement_id": Any<String>,
      "start_date": "2020-03-01T00:00:00.000Z",
    },
    Object {
      "effica_extent_code": "101",
      "effica_extent_nbr": 2,
      "effica_placement_nbr": 13,
      "end_date": "2020-12-31T00:00:00.000Z",
      "id": Any<String>,
      "option_id": "86ef7370-bf85-11eb-91e7-6fcd728c518d",
      "placement_id": Any<String>,
      "start_date": "2020-03-01T00:00:00.000Z",
    },
  ],
  "serviceNeedsTodo": Array [],
}
`;

exports[`GET /transform positive should return transformed special means 1`] = `
Object {
  "assistanceActions": Array [
    Object {
      "actions": Array [
        "10",
      ],
      "child_id": Any<String>,
      "effica_ssn": "010619A9501",
      "end_date": "2020-05-31T00:00:00.000Z",
      "id": Any<String>,
      "start_date": "2019-08-08T00:00:00.000Z",
    },
  ],
  "assistanceActionsTodo": Array [],
}
`;

exports[`GET /transform positive should return transformed special needs 1`] = `
Object {
  "assistanceNeeds": Array [
    Object {
      "bases": Array [
        "DEVELOPMENTAL_DISABILITY",
      ],
      "child_id": Any<String>,
      "effica_ssn": "010619A9501",
      "end_date": "2020-05-31T00:00:00.000Z",
      "id": Any<String>,
      "start_date": "2019-08-08T00:00:00.000Z",
    },
  ],
  "assistanceNeedsTodo": Array [],
}
`;

exports[`GET /transform positive should return transformed timestamps 1`] = `
Object {
  "childAttendances": Array [
    Object {
      "arrived": "2021-01-01T05:26:00.000Z",
      "child_id": Any<String>,
      "departed": "2021-01-01T14:17:00.000Z",
      "effica_rownumber": 1122,
      "id": Any<String>,
      "unit_id": "19fec1fa-e2f1-11eb-8473-eb1f7ce94b07",
    },
  ],
  "childAttendancesTodo": Array [],
}
`;

exports[`GET /transform positive should return transformed voucher value decisions 1`] = `
Object {
  "decisions": Array [
    Object {
      "base_co_payment": 0,
      "capacity_factor": "0.00",
      "child_date_of_birth": "2019-06-01T00:00:00.000Z",
      "child_id": Any<String>,
      "co_payment": 0,
      "decision_number": "1",
      "effica_decision_date": "2020-12-01T00:00:00.000Z",
      "effica_guid": "10",
      "effica_ssn": "010619A9501",
      "family_size": 5,
      "final_co_payment": 100,
      "head_of_family_id": Any<String>,
      "id": Any<String>,
      "partner_id": Any<String>,
      "placement_unit_id": "19fec146-e2f1-11eb-8473-db55258254c5",
      "service_need_option_id": "86ef70a0-bf85-11eb-91e6-1fb57a101161",
      "sibling_discount": 0,
      "status": "SENT",
      "status_group": "APPROVED",
      "valid_from": "2020-04-01T00:00:00.000Z",
      "valid_to": null,
      "voucher_value": 1000,
    },
  ],
  "decisionsTodo": Array [],
}
`;
